src/ralph/models/xapi/profile.py

```diff
@@ -0,0 +1,438 @@
+"""xAPI JSON-LD Profile."""
+
+from datetime import datetime
+from typing import Optional, Union
+
+try:
+    from typing import Literal
+except ImportError:
+    from typing_extensions import Literal
+
+from jsonschema.exceptions import SchemaError
+from jsonschema.validators import validator_for
+from pydantic import AnyUrl, BaseModel, Field, Json, conlist, constr
+
+from ralph.models.selector import selector
+from ralph.models.xapi.base.common import IRI, LanguageMap
+from ralph.models.xapi.base.unnested_objects import (
+    BaseXapiActivityDefinition,
+    BaseXapiActivityInteractionDefinition,
+)
+
+# NB: These pydantic models should follow the profile structure specification.
+# See: https://github.com/adlnet/xapi-profiles/blob/master/xapi-profiles-structure.md
+
+
+class JSONSchema(dict):
+    """Pydantic custom data type validating JSONSchemas."""
+
+    @classmethod
+    def __get_validators__(cls):  # noqa: D105
+        def validate(schema: dict):
+            """Check whether the provided `schema` is a valid JSONSchema."""
+            try:
+                validator_for(schema).check_schema(schema)
+            except SchemaError as error:
+                raise ValueError(f"Invalid JSONSchema: {error}") from error
+            return cls(schema)
+
+        yield validate
+
+
+class ProfilePattern(BaseModel):
+    """Profile `patters` field.
+
+    Note:
+        We do not validate the following requirements:
+        - A `primary` Pattern MUST include `prefLabel` and `definition`.
+          > We accept not including `prefLabel` and `definition`.
+        - A Pattern MUST contain exactly one of `alternates`, `optional`, `oneOrMore`,
+          `sequence`, and `zeroOrMore`.
+          > We accept zero and more than one `alternates`, `optional`, `oneOrMore`,
+          `sequence`, and `zeroOrMore` fields.
+
+    Attributes:
+        id (URI): A URI for the Pattern.
+        type (str): Equal to `Pattern`.
+        primary (bool): Only primary Patterns are checked for matching sequences of
```

Comment:
Couldn't we add a custom validator checking that if the pattern is primary, it must include a `prefLabel` and `definition`?